<a href="https://colab.research.google.com/github/JhonnyLimachi/Sigmoidal/blob/main/78_Introdu%C3%A7%C3%A3o_ao_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---
<img alt="Colaboratory logo" width="15%" src="https://raw.githubusercontent.com/carlosfab/escola-data-science/master/img/novo_logo_bg_claro.png">

#### **Introdução ao SQL**
*by [sigmoidal.ai](https://sigmoidal.ai)*  

---

# Introdução ao SQL

**SQL** significa ***Structured Query Language***. SQL é uma linguagem usada para se comunicar com bancos de dados. De acordo com a ANSI (*American National Standards Institute*), é a linguagem padrão para sistemas de gerenciamento de bancos de dados relacionais.

Através de *statements*, SQL é utilizado para para realizar tarefas como atualizar, retirar ou ler dados em um banco de dados. Alguns Sistemas de Gerenciamento de Banco de Dados que usam SQL são: Oracle, Sybase, Microsoft SQL Server, Microsoft Access, Ingres, etc.

Apesar da maioria dos sistemas de banco de dados usarem SQL, a maioria deles têm alguma extensão proprietária, que geralmente é usada apenas em seu sistema.De qualquer forma, os comandos padrão para SQL como `Select`, `Insert`, `Update`, `Delete`, `Create`, e `Drop`podem ser usados para fazer basicamente qualquer coisa com um banco de dados.

<p align=center>
<img src="https://www.syncfusion.com/blogs/wp-content/uploads/2020/07/Top-10-SQL-Query-Optimization-Techniques.jpg" width="60%"></p>

Não reconheceu algumas das coisas mencionadas acima? Tudo bem! Para quem nunca mexeu com SQL, esse é o material correto! Vamos por partes...

## Nomenclaturas

A internet é, muitas vezes, uma excelente fonte de conhecimento. Mas, em alguns casos, podemos nos deparar com desinformação. Existe muita confusão em relação ao SQL, Bancos de Dados, etc. Por isso, antes de tudo, precisamos alinhar nossas nomenclaturas.

<p align=center><img src="https://images.unsplash.com/photo-1529078155058-5d716f45d604?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1049&q=80" width="60%"></p>

Abaixo, vão algumas das que mais causam confusão:

* **SQL** - Em portugês, chamamos de "Esse-quê-ele", mas em inglês, podemos ver a pronúncia individual de cada letra em inglês, como se estivesse soletrando, ou então como se fosse a palavra em inglês *"sequel"*.
* **Banco de Dados** - Um banco de dados é uma coleção organizada de informação ou dados estruturados armazenados eletronicamente em um computador.
* **Sistema Gerenciador de Banco de Dados** - Um banco de dados normalmente é controlado por um Sistema Gerenciador de Banco de Dados (SGDB ou *DBMS*, em inglês). Juntos, os dados e o SGDB, além de todas as aplicações associadas a eles são referenciados como um sistema de banco de dados, o que é comumente simplificado para a apenas banco de dados.


## Por Que SQL é tão utilizado?

Se você está migrando para a área da tecnologia agora, ou mesmo se você costuma receber seus dados em um arquivo csv, pronto para você usar, talvez não esteja familiarizado com o que é, e a importância do SQL e dos Bancos de Dados e seus Sistemas Gerenciadores nas grandes empresas.

<p align=center><img src="https://www.lifewire.com/thmb/m8HKZDKl4hUMc5BmPx6OaVNwsG8=/2553x1436/smart/filters:no_upscale()/Database-vs-Spreadsheet-6917a9628c7d4f419feefe65acdfb5c0.jpg" width="60%"></p>

Todos nós em algum momento já tivemos que lidar com planilhas. Seja no Google Docs, MS Excel, Numbers, Calc, etc. Mas se esse é um formato muito mais simples e comum entre os "mortais" para armazenar dados, por que não podemos ficar apenas com esses formatos?

Antes de nos aprofundarmos em queries de SQL, vamos dar uma olhada no que faz SQL e Bancos de Dados que utilizam SQL tão populares.

É importante reforçar que SQL é uma linguagem, e até por isso a última palavrava que integra o acrônimo SQL é *Structured Query Language*.

SQL é utilizado no mundo inteiro, e vai muito além do que veremos nesse curso. Com isso em mente, lembre-se que o SQL é reconhecido por causa de sua interaçõ com Bancos de Dados.

Para nós, no momento, podemos pensar em um banco de dados como um conjunto de planilhas, por exemplo, todas armazenadas em um lugar. Entretanto, nem todos os bancos de dados funcionam dessa forma, mas para os propósitos desse curso, podemos entendê-los assim.

Esse tipo de banco de dados onde armazenamos dados estruturados com relacionamento entre as tabelas, onde cada tabela pode ser entendida como uma planilha é o que compõe um Banco de Dados Relacional, com o qual podemos interagir com SQL.

Existem algumas grandes vantagens em usar bancos de dados relacionais tradicionais, interagindo com eles com SQL. Algumas das vantagens são:

* SQL é fácil de entender
* Bancos de Dados Relacionais Tradicionais nos permitem acessar dados diretamente
* Bancos de Dados Tradicionais nos permitem auditar e replicar nossos dados
* SQL é uma excelente ferramenta para analisar várias tabelas de uma vez
* SQL permite que analizemos questões mais complexas do que ferramentas de Dashboard como o Google Analytics


## SQL vs. NoSQL

<p align=center><img src="https://cdn-images-1.medium.com/fit/t/1600/480/0*xoC2DeugTHSh5yFV.jpg" width="60%"></p>

Talvez você já tenha ouvido falar de NoSQL, que significa *Not Only SQL*, ou, em bom português, "Não somente SQL". Bancos de Dados usando NoSQL permitem que você escreva código que interage com os dados de uma forma diferente do SQL que iremos praticar nesse curso.

Os ambientes NoSQL tendem a ser mais populares para dados baseados na web, mas menos popular para dados que vivem em planilhas, da mesma forma dos conjuntos de dados que temos analizados até agora.

Bancos de Dados NoSQL têm *schemas* dinâmicos para dados não-estruturados, e os dados são armazenados em várias formas. Você pode usar abordagens orientadas a colunas, a documentos, a gráficos ou KeyValue. Essa flexibilidade significa que:

* Podemos criar documentos sem ter que definir sua estrutura primeiro
* Cada documento pode ter uma estrutura única e própria
* A sintaxe pode variar a cada banco de dados
* Podemos adicionar campos à medida que formos precisando deles.


## Por Que Negócios Gostam de Bancos de Dados?

<p align=center><img src="https://images.unsplash.com/photo-1427751840561-9852520f8ce8?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1355&q=80" width="50%"></p>

Algumas das coisas que menciomaos anteriorermente ajudam a justificar a popularidade dos Bancos de Dados, e sua interação com SQL, mas vamos a alguns pontos interessantes que merecem destaque.

* **Integridade dos Dados é Grantida** - Apenas os dados que são inseridos compõem o banco de dados, e apenas alguns usuários têm permissão para inserir dados.

* **Os Dados Podem Ser Acessados Rapidamente** - SQL permite que os resultados sejam obtidos a partir do banco de dados de forma muito rápida. O código pode ser otimizado para ter resultados rápidos.

* **Os Dados São Facilmente Compartilhados** - Vários indivíduos podem acessar os dados armazenados no banco de dados, e os dados são os mesmos para todos, permitindo resultados consistentes para todos que acessam o banco de dados.


## Como um Banco Relacional Fucniona?

<p align=center><img src="https://www.researchgate.net/profile/Radoslava-Kraleva/publication/323466947/figure/fig1/AS:599227063353344@1519878273386/Structure-of-the-relational-database-Mvsemdm-Each-box-on-the-figure-above-contains-one.png" width="50%"></p>

Em bancos de dados relacionais os dados são armazenados em tabelas, que podem ser entendidas como planilhas de Excel, por exemplo.

Na maior parte do tempo, podemos pensar em bancos de dados como um conjunto de planilhas de Excel, cada planilha com linhas e colunas, onde cada linha contém dados sobre uma transação, uma pessoa, uma empresa, etc. As colunas, contém dados sobre um aspecto de cada uma das linhas como nomes, local, número de identificação, etc.

**Todos as entradas em uma mesma coluna devem ter o mesmo tipo de dado.** A coluna inteira é considerada quantitativa, discreta ou algum tipo de string. Isso significa que se você tiver uma string em uma linha de uma determinada coluna, a coluna inteira pode mudar para string. Isso pode ser muito complicado caso você deseje fazer operações matemáticas com essa coluna.

Estruturas de dados consistentes entre as colunas é uma das principais razões pelas quais bancos de dados são tão rápidos. É muito comum que bancos de dados armazenem uma enorme quantidade de dados, e por isso, essa propriedade é muito importante, pois garante integridade dos dados e agilidade nas consultas e coletas.

## Entity Relationship Diagrams (Diagrama Entidade-Relacionamento)

<p align=center><img src="https://www.impacttechnology.co.uk/media/images/Example-ERD-800px.png" width="60%"></p>



Um ERD é uma forma comum de visualizar os dados em um banco de dados. Podemos visualizar, por exemplo:

* Os nomes das tabelas.
* As colunas em cada tabela.
* A forma de interação dessas colunas.
* Podemos entender cada uma das tabelas como uma "planilha".

Abaixo, podemos ver um pouco mais sobre esse tipo de diagrama. A fonte do trecho abaixo que explica em detalhe esse relacionamento pode ser encontrado [aqui](https://www.lucidchart.com/pages/pt/o-que-e-diagrama-entidade-relacionamento).

> Um diagrama entidade relacionamento (ER) é um tipo de fluxograma que ilustra como “entidades”, p. ex., pessoas, objetos ou conceitos, se relacionam entre si dentro de um sistema. Diagramas ER são mais utilizados para projetar ou depurar bancos de dados relacionais nas áreas de engenharia de software, sistemas de informações empresariais, educação e pesquisa. Também conhecidos como DERs, ou modelos ER, usam um conjunto definido de símbolos, tais como retângulos, diamantes, ovais e linhas de conexão para representar a interconectividade de entidades, relacionamentos e seus atributos. Eles espelham estruturas gramaticais, onde entidades são substantivos e relacionamentos são verbos.
>
> Diagramas ER estão relacionados com diagramas de estrutura de dados (DEDs), que incidem sobre as relações de elementos dentro de entidades em vez de relações entre as próprias entidades. Diagramas ER são também muitas vezes utilizados junto com diagramas de fluxo de dados (DFDs), que mapeiam o fluxo de informações para processos ou sistemas.
>
> ### Componentes e funcionalidades de um diagrama ER
>
> Diagramas ER são compostos de entidades, relacionamentos e atributos. Eles também descrevem a cardinalidade, que define as relações em termos de números. Veja um glossário:
>
> ### Entidade
>
>Algo que pode ser definido e que pode ter dados armazenados sobre ele — como uma pessoa, um objeto, conceito ou evento. Pense em entidades como substantivos. Exemplos: um cliente, estudante, carro ou produto. Normalmente representado como um retângulo.
>
>
>**Tipo de entidade**: um grupo de coisas definíveis, como estudantes ou atletas, ao passo que a entidade seria um estudante ou atleta específico. Outros exemplos: clientes, carros ou produtos.
>
>**Conjunto de entidade**: assim como um tipo de entidade, mas definido em um determinado ponto no tempo, tal como estudantes matriculados em um curso no primeiro dia de aula. Outros exemplos: clientes que fizeram compras no mês passado, carros atualmente registrados na Flórida. Um termo relacionado é ‘instância’, em que a pessoa ou carro específico seria uma instância do conjunto de entidade.
>
>**Categorias de entidades**: entidades são categorizadas como forte, fraca ou associativa. Uma entidade forte pode ser definida unicamente pelos seus próprios atributos, enquanto uma entidade fraca não pode. Uma entidade associativa junta entidades (ou elementos) dentro de um conjunto de entidades.
>
>
>**Chaves de entidades**: refere-se a um atributo que define exclusivamente uma entidade dentro de um conjunto de entidades. Chaves de entidades podem ser super, candidata ou primária. Superchave: um conjunto de atributos (um ou mais) que, juntos, definem uma entidade em um conjunto de entidades. Chave candidata: uma superchave mínima, o que significa que tem o menor número possível de atributos para ser uma superchave. Um conjunto de entidades pode ter mais de uma chave candidata. Chave primária: uma chave candidata escolhida pelo designer do banco de dados para identificar exclusivamente o conjunto de entidades. Chave estrangeira: identifica o relacionamento entre entidades.
>
> ### Relacionamento
>
>Como entidades atuam umas sobre as outras ou estão associadas uma com a outra. Pense em relacionamentos como verbos. Por exemplo, o estudante pode se inscrever em um curso. As duas entidades seriam o aluno e o curso, e o relacionamento descrito é o ato de matricular-se, assim conectando as duas entidades. Relacionamentos são tipicamente representados por diamantes ou etiquetas diretamente sobre as linhas de ligação.
>
>
>**Relacionamento recursivo**: a mesma entidade participa mais de uma vez na relação.
>
> ### Atributo
>
>A propriedade ou característica de uma entidade, muitas vezes representada por um oval ou círculo.
>
>
>**Atributo descritivo**: uma propriedade ou característica de uma relação (versus de uma entidade).
>
>**Categorias de atributos**: atributos são classificados como simples, composto, derivado, bem como de valor único ou multivalor. Simples: significa que o valor do atributo é atômico e não pode ser dividido, como um número de telefone. Composto: subatributos nascem de um atributo. Derivado: o atributo é calculado, ou derivado de outro atributo, tal como a idade a partir de uma data de nascimento.
>
>
>**Multivalor**: mais de um valor de atributo é identificado, como vários números de telefone para uma única pessoa.
>
>
>**Valor único**: apenas um valor de atributo. Os tipos podem ser combinados, tais como: atributos de valor único e simples ou atributos de multivalor compostos.
>
>### Cardinalidade
>
>Define os atributos numéricos da relação entre duas entidades ou conjuntos de entidades. Os três principais relacionamentos cardinais são um-para-um, um-para-muitos e muitos-para-muitos. Um exemplo de um-para-um seria um estudante associado a um endereço de correspondência. Um exemplo de um-para-muitos (ou muitos-para-um, dependendo do sentido da relação): um estudante se inscreve para vários cursos, mas todos esses cursos têm uma única linha que leva de volta ao aluno. Exemplo de muitos-para-muitos: estudantes como um grupo são associados a vários membros do corpo docente, e membros do corpo docente, por sua vez, são associados a vários alunos.
>
>**Visualizações de cardinalidade**: a cardinalidade pode ser representada por interpolação ou do mesmo lado, dependendo de onde os símbolos são mostrados.
>
>**Restrições de cardinalidade**: os números mínimos ou máximos que se aplicam a um relacionamento.



## Statements

Nós utilizamos a linguagem SQL para interagir com bancos de dados através do que chamamos de *statements*, ou sentenças.

Aqui, vamos olhar os tipos de *statements* que podemos utilizar.

<p align=center><img src="https://images.unsplash.com/photo-1554306274-f23873d9a26c?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3150&q=80" width="70%"></p>


### Tipos de SQL Statements

Abaixo, vemos um resumo de statements SQL divididos nas seguintes categorias:

* Data Definition Language (DDL) Statements
* Data Manipulation Language (DML) Statements
* Transaction Control Statements
* Session Control Statements
* System Control Statement
* Embedded SQL Statements
* Data Definition Language (DDL) Statements


Com esses tipos de Statements em mente, podemos analisar cada um deles, e suas funções

### Data Definition Language

Os statements dessa categoria são utilizados para:

* Criar, alterar e remover objetos do tipo esquema (*schema*)
* Conceder e revogar privilégios e papéis para colaboradores
* Analisar informações de tabelas, índices ou clusters
* Estabalecer opções para auditoria
* Adicionar comentários aos dicionários de dados

Os comandos `CREATE`, `ALTER` e `DROP` requerem acesso exclusivo ao objeto alvo. Por exemplo, um statement `ALTER_TABLE` irá falhar caso outro usuário esteja com aquela tabela específica aberta no momento.

Os comandos `GRANT`, `REVOKE`, `ANALYZE`, `AUDIT`, e `COMMENT` não requerem acesso exclusivo ao alvo. Por exemplo, é possível analizar uma tabela enquanto outros usuários a atualizam.


`ALTER` ... (All statements beginning with ALTER)

`ANALYZE`

`ASSOCIATE STATISTICS`

`AUDIT`

`COMMENT`

`CREATE` ... (All statements beginning with CREATE)

`DISASSOCIATE STATISTICS`

`DROP` ... (All statements beginning with DROP)

`FLASHBACK` ... (All statements beginning with FLASHBACK)

`GRANT`

`NOAUDIT`

`PURGE`

`RENAME`

`REVOKE`

`TRUNCATE`

## Data Manipulation Language (DML) Statements

Statements do tipo Data Manipulation Language acessam e manipulam os dados nos objetos do *schema* existente. Esses statements são:

`CALL`

`DELETE`

`EXPLAIN PLAN`

`INSERT`

`LOCK TABLE`

`MERGE`

`SELECT`

`UPDATE`

O *statement* `SELECT` é uma forma limitada de DML, pois ele só consegue acessar dados no banco de dados. Ele não pode manipular os dados no banco de dados, apesar de poder operar nos dados acessados antes de retornar resultados na *query*.

## Transaction Control Statements
Transaction Control Statements gerenciam mudanças feitas pelos *statements* DML. Eles são:

`COMMIT`

`ROLLBACK`

`SAVEPOINT`

`SET TRANSACTION`

## Cheat Sheet

É muita coisa e nós sabemos. Lembrar de tudo isso de cara é tarefa difícil, e precisamos entender como usar cada um desses tipos de statements individualmente. Isso é o que faremos quando chegarmos à nossa parte mais prática desse curso.

<p align=center><img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSLC_5u1i2wwUA_qp0tHRk6k8cIXrytlwO5pA&usqp=CAU" width="60%"></p>


Por enquanto, vamos deixar vocês com esse [Cheat Sheet](https://www.dataquest.io/blog/sql-cheat-sheet/) de SQL para Data Science, onde você consegue ter uma referência, com exemplos práticos, em código, de *statements* SQL.

## Os Dados

Para o nosso projeto, utilizaremos uma versão reduzida do nosso conjunto de dados do Airbnb, do projeto 1 do Data Science na Prática.

<p align=center><img src="https://images.unsplash.com/photo-1587161584760-f51779fb276a?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1050&q=80" width="60%"></p>

Utilizaremos dados de *New York City*, mas com algumas colunas a menos, e com apenas uma fração das linhas. O conjunto de dados foi disponibilizado junto dos links da aula, mas pode ser baixado também por [aqui](https://raw.githubusercontent.com/rafaelnduarte/sigmoidal_data/master/data.csv)

Vamos visualizar esse conjunto para sabermos melhor com o que estamos lidando.

In [ ]:
# importando pacotes e dados

import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/rafaelnduarte/sigmoidal_data/master/data.csv', index_col='id')

print(df.shape)
df.head()

(5509, 9)


name    host_id  \
id                                                                       
46177828  Amazing 2 bedroom apt with  Privet Garden & Of...  158092208   
5305619                 Newly Renovated Flat, McKibbin Loft    5663328   
33497909          Beautiful, Comfortable Studio In Midtown!  252432969   
10625668                                         The Orchid   26354652   
51288148  24/7 Crashpad Centrally Located NextDoor to Su...  262459086   

         host_name neighbourhood_group neighbourhood        room_type  price  \
id                                                                             
46177828   Evandra           Manhattan        Harlem  Entire home/apt    230   
5305619      Scott            Brooklyn  Williamsburg     Private room     75   
33497909     Jaime           Manhattan       Midtown  Entire home/apt    260   
10625668  Philbert            Brooklyn      Flatbush     Private room    126   
51288148      Kirt            Brooklyn  Clinton Hill      Shared room     38   

          minimum_nights  number_of_reviews  
id                                           
46177828               5                  4  
5305619               30                  0  
33497909              30                 82  
10625668               2                  1  
51288148               5                  0

Podemos verificar que nosso conjunto de dados possui 5509 linhas, para 9 colunas.

O conjunto original possuía mais colunas, mas que não eram interessantes para a nossa análise. Aqui, o foco é entender como funcionam bancos de dados, e como utilizar de *queries* em SQL para acessar esses conjuntos de dados.

Abaixo, podemos ver os tipos dos dados, e também se existem dados ausentes.

In [ ]:
# distribuição dos dados
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5509 entries, 46177828 to 23488488
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   name                 5506 non-null   object
 1   host_id              5509 non-null   int64 
 2   host_name            5506 non-null   object
 3   neighbourhood_group  5509 non-null   object
 4   neighbourhood        5509 non-null   object
 5   room_type            5509 non-null   object
 6   price                5509 non-null   int64 
 7   minimum_nights       5509 non-null   int64 
 8   number_of_reviews    5509 non-null   int64 
dtypes: int64(4), object(5)
memory usage: 430.4+ KB


## A Plataforma
<p align=center><img src="https://images.unsplash.com/photo-1544652478-6653e09f18a2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1050&q=80" width="60%"></p>

Para que possamos passar mais tempo focando nas *queries* e na linguagem em si, iremos utilizar o [SQLite Online](https://sqliteonline.com), um ambiente online, onde não precisaremos nos preocupar com o setup de um servidor, e estaremos expostos a um ambiente mais próximo do que costuma ser a realidade dos cientista de dados, onde o foco é o acesso aos dados e não o setup, criação e manutenção dos bancos de dados.

Com isso em mente, basta fazer o upload dos dados que disponibilizamos na plataforma, e começar a criar suas *queries*.

Ao fazer o upload dos dados utilizando o botão `import` no menu superior da plataforma, lembre-se de definir o nome das colunas como "primeira linha".

Em nosso arquivo csv, a primeira linha é o nome das colunas. Para que a nossa tabela seja criada da forma correta, é imprescindível que esse parâmetro seja configurado da forma correta.

Com isso feito, nossos dados estão prontos para serem consumidos! Mãos à obra!


## Boas Práticas

* Letras Maiúsculas

Você pode notar que sempre colocamos as sentenças `SELECT` e `FROM`, enquantoo deixamos os nomes das tabelas e colunas em letras minúsculas. Isso é uma convenção de formatação dentro do SQL, e você irá reparar isso na maioria, se não todos, os lugares que você for pegar referências de SQL.

Essa prática de deixar os comandos em letras maiúsculas e o restante da query em letras minúsculas faz com que as queries sejam mais fáceis de ler, o que vai ser muito importante quando você começar a ler e escrever queries mais complexas.

Por enquanto, é só uma boa prática que você já pode ir se habituando a fazer!

* Evite Espaços em Tabelas e Nomes de Variáveis

Assim como já costumamos fazer quando estamos manipulando datasets no pandas aqui no Jupyter Notebook, dê preferência por usar underline (_) e evite deixar espaços em brancos nos nomes de suas tabelas e variáveis.

Trabalhar com esses espaços pode ser um pouco incômodo com SQL. Por exemplo, no Postgres, se você tiver espaços entre os nomes de colunas ou tabelas, você precisa se referir a esses elementos com aspas duplas entre eles.

`SELECT * FROM "Vendas Agosto"`

Ao invés do bom e velho:

`SELECT * FROM vendas_agosto`

Na maioria dos ambientes, você pode ver chaves ao invés de aspas duplas, como por exemplo:

`SELECT * FROM [Vendas Agosto]`

* Use espaços nas Queries

Queries em SQL ignoram espaços, portanto, podemos adicionar quantos espaços e até linhas em branco quisermos, e isso não afetará as queries.

Por exempolo:

`SELECT id FROM vendas_agosto`

É equivalente a:

`SELECT id`

`FROM vendas_agosto`

Mas é claro, se utilize disso para fazer queries mais legíveis e fáceis de entender, mas não exagere nos espaços em branco.

* SQL não é *Case Sensitive*

Em Python, sabemos que `df` é diferente de `Df`, que é diferente de `DF`. Isso acontece pois Python é *case sensitive*, ou seja, diferencia maiúsculas de minúsculas. O mesmo não é verdade para SQL.

As seguintes queries trariam os mesmos resultados:

`SELECT id`

`FROM vendas_agosto`

É igual a:

`select id`

`from vendas_agosto`

Que é  igual a:

`SeLeCt iD`

`FrOm VeNdAs_AgOsTo`

Entretanto, mais uma vez, o ideal é que você se mantenha ao padrão que mostramos anteriormente.


* Ponto e Vírgula (;)
A depender do ambiente de SQL que você esteja utilizando, a sua query pode precisar de ponto e vírgula ao final para ser executada.

Apesar disso, nem todos os ambientes são assim, com alguns sendo mais flexíveis com essa regra. Entretanto, é considerada uma boa prática adicionar ponto e vírgula ao final de suas queries.

Exemplo, completo, utilizando das boas práticas mencionadas até aqui:

`SELECT id`

`FROM vendas_agosto;`


## FROM

O comando FROM irá indicar, como o próprio nome sugere, de onde iremos tirar os dados. Ele se refere à Tabela de origem dos dados.

Por exemplo, vamos supor que temos as tabelas:

`vendas`, `funcionarios`, `estoque`

e queremos selecionar todos os dados da tabela `vendas`.

Nossa query seria:

`SELECT *`

`FROM vendas`

Para puxar dados de uma tabela diferente, basta mudar a tabela:

`SELECT *`

`FROM funcionarios`


## WHERE

O comando `WHERE` vai ser utilizado para fazer um filtro, onde os dados que serão puxados deverão obedecer à condição estabelecida por ele.

Por exemplo:

`SELECT * FROM data`

`WHERE neighbourhood_group = 'Manhattan'`

Acima, como nossa condição é um String, percebemos que precisamos utilizar as aspas. Em alguns sistemas, aspas duplas também será possíveis.

Quando a condição é um valor numérico, não utilizaremos as aspas.

`SELECT * FROM data`

`WHERE minimum_nights = 2`

Abaixo, podemos ver uma lista de operadores que podemos utilizar quando estivermos trabalhando com o `WHERE`:

* `=` - Igual
* `>` - Maior
* `<` - Menor
* `>=` - Maior ou Igual
* `<=` - Menor ou Igual
* `<>` - Não igual (Em algumas sistemas podemos utilizar `!=`)
* `BETWEEN` - Entre um certo range
* `LIKE` - Busca por um padrão
* `IN` - Especifica vários possíveis valores dentro de uma coluna

## ORDER BY

Como o próprio nome sugere, o `ORDER BY` é usado para ordenar os resultados da nossa query. Por padrão, os resultados são colocados em ordem ascendente, mas podem ser colocados em ordem descendente, usando o comando `DESC`.

Query de exemplo:

`SELECT *`

`FROM data`

`ORDER BY price DESC`

## LIMIT

Mais uma vez, lembramos o quanto o SQL pode ser literal. O `LIMIT` nada mais é do que um limite. Podemos limitar o número de linhas que vamos receber da nossa Query.

Vamos supor que queremos apenas 10 entradas do nosso conjunto de imóveis em Manhattan, podemos fazer a seguinte query:

`SELECT * FROM data`

`WHERE neighbourhood_group = 'Manhattan'`

`LIMIT 10;`

## Operadores Aritméticos

O SQL também nos permite trabalhar com operadores aritméticos. Aqui, vamos utilizar dos mesmos que já conhecemos em Python.

* `+` - Adição
* `-` - Subtração
* `*` - Multiplicação
* `/` - Divisão
* `%` - Modulo

Podemos fazer uso desses operadores tanto no `SELECT`, para que retorne uma nova coluna de consulta com a nossa operação, quanto no `WHERE` para satisfazer determinada condição. Também podemos fazer as duas coisas.

## Operadores Lógicos

Mais uma vez, assim como no Python, podemos utilizar operadores lógicos em nossas queries para diferentes funções.

Segue abaixo uma lista dos operadores lógicos e suas funções:

* `ALL` - True se todos os valores da query satisfazem a condição

Pode ser usada, por exemplo, para puxar todas as entradas de uma coluna que estão preenchidas.

`SELECT ALL name`

`FROM data`

`WHERE TRUE;`


* `AND` - True se todas as condições separadas por AND são verdadeiras

A ideia aqui é usar dentro do WHERE mais de uma condição a ser obedecida.

`SELECT * FROM data`

`WHERE neighbourhood_group = 'Manhattan' AND price < 100;`


* `OR` - True se alguma das condições separadas por OR são verdadeiras

Da mesma formas que colocamos condições com o OR no WHERE, podemos utilizar o OR.

`SELECT * FROM data`

`WHERE price <100 AND neighbourhood_group = 'Manhattan' OR neighbourhood_group= 'Brooklyn';`

* `NOT` - Mostra regristros caso a condição não seja verdadeira

Já o NOT, podemos utilizar para excluir registros de nossa busca.

`SELECT * FROM data`

`WHERE NOT neighbourhood_group = 'Manhattan' AND price < 100;`

* `IN` - True se o operador está dentro da lista de expressões

O IN nos permite buscar vários valores dentro de um WHERE. Pode ser um atalho para utilizar vários OR

`SELECT * FROM data`

`WHERE neighbourhood IN ('Astoria', 'Long Island City', 'Flatbush', 'Jackson Heights');`

Também poderíamos acrescentar um NOT a esse comando para excluir entradas com esses valores.

* `BETWEEN` - True se o operador estiver dentro do range de comparação

`SELECT * FROM data`

`WHERE price BETWEEN 100 and 200;`

* `LIKE` - True se o operador é compatível a um padrão
Com esse operador conseguimos investigar mais a fundo nosso conjunto de dados. Esse operador normalmente é utilizado junto do sinal `%` e `_`. Por exemplo, para retornar todos os donos de anúncios que começam com a letra A:

`SELECT * FROM data`

`WHERE host_name LIKE 'A%'`

## **Extra: O que é SQL e por que preciso saber disso?**

SQL significa Structured Query Language. SQL é uma linguagem usada para se comunicar com bancos de dados. De acordo com a ANSI (American National Standards Institute), é a linguagem padrão para sistemas de gerenciamento de bancos de dados relacionais.

Através de statements, SQL é utilizado para para realizar tarefas como atualizar, retirar ou ler dados em um banco de dados. Alguns Sistemas de Gerenciamento de Banco de Dados que usam SQL são: Oracle, Sybase, Microsoft SQL Server, Microsoft Access, Ingres, etc.

Apesar da maioria dos sistemas de banco de dados usarem SQL, a maioria deles têm alguma extensão proprietária, que geralmente é usada apenas em seu sistema.De qualquer forma, os comandos padrão para SQL como Select, Insert, Update, Delete, Create, e Droppodem ser usados para fazer basicamente qualquer coisa com um banco de dados.

Links úteis:

[SQL Cheat Sheet](https://www.dataquest.io/blog/sql-cheat-sheet/)

[Entity Relationship Diagram](https://www.lucidchart.com/pages/er-diagrams)

[MER e DER - DevMedia](https://www.devmedia.com.br/mer-e-der-modelagem-de-bancos-de-dados/14332)

[Como fazer um Diagrama Entidade Relacionamento](https://www.lucidchart.com/pages/pt/como-fazer-um-diagrama-entidade-relacionamento)

[Entenda o Modelo Entidade Relacionamento](https://www.estrategiaconcursos.com.br/blog/modelo-entidade-relacionamento/)

## **Extra 2 : SQL vs NoSQL**

[NoSQL Cheat Sheet for Dummies](https://www.dummies.com/category/articles/big-data-33578/)

[O que são Bancos de Dados NoSQL? - Microsoft Azure](https://azure.microsoft.com/pt-br/resources/cloud-computing-dictionary/what-is-nosql-database/)

[What is NoSQL - MongoDB](https://www.mongodb.com/resources/basics/databases/nosql-explained)

[Introdução aos Bancos de Dados NoSQL - DevMedia](https://www.devmedia.com.br/introducao-aos-bancos-de-dados-nosql/26044)

## **Relacionamentos entre tabelas**

Links úteis:

[SQL Cheat Sheet](https://www.dataquest.io/blog/sql-cheat-sheet/)

[Entity Relationship Diagram](https://www.lucidchart.com/pages/er-diagrams)

[MER e DER - DevMedia](https://www.devmedia.com.br/mer-e-der-modelagem-de-bancos-de-dados/14332)

[Como fazer um Diagrama Entidade Relacionamento](https://www.lucidchart.com/pages/pt/como-fazer-um-diagrama-entidade-relacionamento)

[Entenda o Modelo Entidade Relacionamento](https://www.estrategiaconcursos.com.br/blog/modelo-entidade-relacionamento/)

## **Extra 3: SQL Online IDE**

Nessa aula, aprendemos a subir os dados para a nossa IDE, e configurar nosso ambiente.

Links úteis:

[SQL Online IDE](https://insideairbnb.com/get-the-data/)

[Inside Airbnb - Dados Originais](https://insideairbnb.com/get-the-data/)


## **Extra 4: Boas Práticas em SQL**

Links úteis:

[SQL Online IDE](https://insideairbnb.com/get-the-data/)

[Inside Airbnb - Dados Originais](https://insideairbnb.com/get-the-data/)


## **Extra 5: SELECT e FROM**

Links úteis:

[SQL Online IDE](https://insideairbnb.com/get-the-data/)

[Inside Airbnb - Dados Originais](https://insideairbnb.com/get-the-data/)



## **Extra 6: WHERE**

Links úteis:

SQL Online IDE

Inside Airbnb - Dados Originais
